### ISSUE TITLE
Sweep is not running on products that have been restored

### ISSUE THREAD

**Issue by hsaraogi** [palantir/atlasdb#2807]
issue: We have a fix #2778, but products will have to release with the fix and impls to upgrade to the latest/patch versions, this would be a very tight schedule before the break.

Proposed solution: We should have a CLI that can fix up the issue on the affected deployment.

**Option1**
There is practically no way of determining from the`_transactions` table alone as to which timestamps were deleted by the CLI once the CLI has done its job.. We might be able to guess using the restore timestamp if the impl remembers the restore timestamp for each product. The service will have to be offline.

```
for (i = restoreTimestamp; ;i++) {
       Long commitTs = read commitTs from "_transactions" where startTs = i;
       if (commitTs == null) {
                  write (startTs=i, commitTs = -1) to "_transactions";
       } else {
                  break;
       }
}
```

Con:
Lots of redundant values in the transactions table, which will never be cleaned up as a lot of the timestamps will not be startTs at all.

**Option2**
If we transactionally read entire tables, i.e. a rangescan.ALL for all tables, we can cause all uncommitted transactions to rollback.

Con:
1. This CLI will be slow.
2. It can cause heavy load on Cassandra.
3. This will have to be schema aware.

**Option3**:

Run the migration CLI and skip writes. The impl will just involve adding a configuration option to the migration CLI to skip writes?

**Comment by hsaraogi** [comment]
**Option3** is the lowest lift :)

**Comment by hsaraogi** [comment]
Option4 from @nziebart:
We can compact the transactions table. I think we just need to `alter table _transactions set gc_grace_seconds = 0` and then `nodetool flush && nodetool compact <keyspace> _transactions`

**Comment by tpetracca** [comment]
Briefly talked offline with Himangi, but +1 to option 4

I don't believe the `gc_grace_seconds` altering is required given we're already running with it at an hour.

Only potential downfall is if we're in a repaired vs. unrepaired situation where the tombstones don't get cleaned up. If that's the case we're already in a bad state and are automating this internally in a release this week.

**Comment by hsaraogi** [comment]
tesh.sh:
```
while read lineOfKeyspaces; do
    for keyspace in $lineOfKeyspaces; do
        echo "keyspace = '$keyspace'"
        if [[ $keyspace == *"system"* ]]; then
                echo "Skipping as this is a system keyspace!"
                continue;
        fi
        txTable=" _transactions"
        echo $txTable
        ccm node1 cqlsh -e "use $keyspace; alter table \"_transactions\" with gc_grace_seconds = 0;"
        ccm node1 nodetool flush && ccm node1 nodetool compact $keyspace$txTable
        echo "Compacted the $txTable in keyspace $keyspace!"
    done
done

```

Confirmed that `ccm node1 cqlsh -e "describe keyspaces;" | ./test.sh` would let a failing sweep progress later on.

After Running the clean transactions CLI:

```
INFO  [2017-12-12 02:13:19,193] com.palantir.atlasdb.sweep.SweepTaskRunner: Beginning iteration of sweep for table default.todo starting at row 0007ceb4e0fce3f0
INFO  [2017-12-12 02:13:19,247] com.palantir.atlasdb.sweep.SpecificTableSweeper: Failed to sweep table default.todo at row 0007ceb4e0fce3f0 with candidate batch size 1, delete batch size 1, and 1 cell+timestamp pairs to examine.
! java.lang.NullPointerException: null
....
```

After Running `ccm node1 cqlsh -e "describe keyspaces;" | ./test.sh`:

```
127.0.0.1 - - [12/Dec/2017:02:14:51 +0000] "POST /todos/ HTTP/1.1" 204 - "-" "unknown-atlasdb (unknown)" 40
127.0.0.1 - - [12/Dec/2017:02:15:01 +0000] "POST /todos/ HTTP/1.1" 204 - "-" "unknown-atlasdb (unknown)" 16
127.0.0.1 - - [12/Dec/2017:02:15:11 +0000] "POST /todos/ HTTP/1.1" 204 - "-" "unknown-atlasdb (unknown)" 16
INFO  [2017-12-12 02:15:19,266] com.palantir.atlasdb.sweep.SweepTaskRunner: Beginning iteration of sweep for table default.todo starting at row 0007ceb4e0fce3f0
INFO  [2017-12-12 02:15:19,413] com.palantir.atlasdb.persistentlock.LockStore: Successfully acquired the persistent lock: LockEntry{lockName=BackupLock, instanceId=812c4f33-1133-4f56-9022-f444acaccf1f, reason=Sweep}
INFO  [2017-12-12 02:15:19,417] com.palantir.atlasdb.sweep.PersistentLockManager: Successfully acquired persistent lock for sweep: PersistentLockId{value=812c4f33-1133-4f56-9022-f444acaccf1f}
INFO  [2017-12-12 02:15:19,420] com.palantir.atlasdb.sweep.PersistentLockManager: Releasing persistent lock PersistentLockId{value=812c4f33-1133-4f56-9022-f444acaccf1f}
INFO  [2017-12-12 02:15:19,438] com.palantir.atlasdb.persistentlock.LockStore: Successfully released the persistent lock: LockEntry{lockName=BackupLock, instanceId=812c4f33-1133-4f56-9022-f444acaccf1f, reason=Sweep}
INFO  [2017-12-12 02:15:19,438] com.palantir.atlasdb.sweep.SpecificTableSweeper: Analyzed 1 cell+timestamp pairs from table default.todo starting at row 0007ceb4e0fce3f0 and deleted 1 stale values in 180 ms up to timestamp 6010534.
```

restarted the stack and see sweep start failing with batch size of 128 because compaction was still running, sweep started going through after a few backoffs though.

**Comment by tpetracca** [comment]
FYSA this is the command I've been running from the service home directory of each Cassandra node:
```
find ./mnt/cassandra-data-0/cassandra-data/*/_transactions-* -type d | cut -d/ -f5 | sort -u | xargs -n1 -I{} ./service/bin/nodetool compact {} _transactions
```
Where per internal tooling and default setups, the data directory location is `<service-home>/mnt/cassandra-data-0/cassandra-data`

**Comment by hsaraogi** [comment]
Send out an email for wider comms.

**Comment by hsaraogi** [comment]
Closing out as we have a fix to make sweep proceed after a restore. Also, we are tracking the transactions CLI fixes elsewhere #2779 and #2861.
